src/infra/core/src/compact_service_impl.rs

```diff
@@ -0,0 +1,289 @@
+        Ok(compact_dir_path)
+    }
+
+    async fn replace_data_files(
+        &self,
+        dataset: Arc<dyn Dataset>,
+        new_file_path: &Path,
+        block_hash_to_reset: &Multihash,
+        last_dataslice_info: &LastDatasliceInfo,
+    ) -> Result<(), CompactError> {
+        let chain = dataset.as_metadata_chain();
+        chain
+            .set_ref(
+                &BlockRef::Head,
+                block_hash_to_reset,
+                SetRefOpts {
+                    validate_block_present: true,
+                    check_ref_is: None,
+                },
+            )
+            .await?;
+
+        let add_data_params = AddDataParams {
+            prev_checkpoint: None,
+            prev_offset: last_dataslice_info.prev_offset,
+            new_offset_interval: last_dataslice_info.new_offset_interval.clone(),
+            new_watermark: None,
+            new_source_state: None,
+        };
+
+        dataset
+            .commit_add_data(
+                add_data_params,
+                Some(OwnedFile::new(new_file_path)),
+                None,
+                CommitOpts {
+                    block_ref: &BlockRef::Head,
+                    system_time: Some(Utc::now()),
+                    prev_block_hash: Some(Some(block_hash_to_reset)),
+                    check_object_refs: false,
+                },
+            )
+            .await?;
+
+        Ok(())
+    }
+
+    fn remove_old_files(
+        &self,
+        chain_files_info: &ChainFilesInfo,
+        listener: &Arc<dyn CompactionListener>,
+    ) -> Result<(), CompactError> {
+        listener.begin_phase(CompactionPhase::CleanDatasliceFiles);
+        for file_url in &chain_files_info.data_slice_file_urls {
+            fs::remove_file(file_url.replace("file://", "")).int_err()?;
+        }
+
+        listener.begin_phase(CompactionPhase::CleanBlockFiles);
+        let block_hash_to_reset_string = chain_files_info
+            .block_hash_to_reset
+            .as_multibase()
+            .to_string();
+        for block_file_url in &chain_files_info.block_file_urls {
+            if block_file_url.contains(&block_hash_to_reset_string) {
+                break;
+            }
+            fs::remove_file(block_file_url.replace("file://", "")).int_err()?;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl CompactService for CompactServiceImpl {
+    #[tracing::instrument(level = "info", skip_all)]
+    async fn compact_dataset(
+        &self,
+        dataset_handle: &DatasetHandle,
+        dataset_dir_path: &Path,
+        multi_listener: Option<Arc<dyn CompactionMultiListener>>,
+    ) -> Result<(), CompactError> {
+        let listener = multi_listener
+            .and_then(|l| l.begin_compact(dataset_handle))
+            .unwrap_or(Arc::new(NullCompactionListener {}));
+        self.dataset_authorizer
+            .check_action_allowed(dataset_handle, domain::auth::DatasetAction::Read)
+            .await?;
```

Comment:
Compaction is both read and write. Checking `Write` action should work.
Right now you have allowed anyone to compact public datasets.

Reply:
Oh, you are right, it was by mistake. In my mind it should be `Write` permission

Reply:
Changed